yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts

```diff
@@ -0,0 +1,210 @@
+import { toArray } from '@aztec/foundation/iterable';
+import type { AztecAsyncKVStore, AztecAsyncMap } from '@aztec/kv-store';
+import type { DirectionalAppTaggingSecret, PreTag } from '@aztec/stdlib/logs';
+import { TxHash } from '@aztec/stdlib/tx';
+
+/**
+ * Data provider of tagging data used when syncing the sender tagging indexes. The recipient alternative of this class
+ * is called RecipientTaggingDataProvider. We have the providers separate for the sender and recipient because
+ * the algorithms are completely disjoint and there is not data reuse between the 2.
+ */
+export class SenderTaggingDataProvider {
+  #store: AztecAsyncKVStore;
+
+  // Stores the pending indexes for each directional app tagging secret. Pending here means that the tx that
+  // contained the private logs with tags corresponding to these indexes has not been finalized yet. We don't store
+  // just the highest index because it could happen that some of the transactions are dropped and then we need
+  // the information about the lower pending indexes. But we store just one index per secret-txHash pair as only
+  // the highest index in the given tx is relevant for future index choice when sending a private log.
```

Comment:
```suggestion
  // Stores the pending indexes for each directional app tagging secret. Pending here means that the tx that
  // contained the private logs with tags corresponding to these indexes has not been finalized yet.
  //
  // We don't store just the highest index because if their transaction is dropped we'd then need the
  // information about the lower pending indexes. For each secret-tx pair however we only store the largest index used in that tx, since the smaller ones are irrelevant due to tx atomicity (this assumes no
  // logs are used in the non-revertible phase).
```

>this assumes no logs are used in the non-revertible phase

I think this is true?

Reply:
Yes, your changes improved it 👍 

Good point on the non-revertible phase logs. That seems like a problem we need to tackle. Would create an issue and link it here, ok?

Reply:
Update: Just noticed I already have an [issue](https://github.com/AztecProtocol/aztec-packages/issues/17615) for dropping the non-revertible phase logs.